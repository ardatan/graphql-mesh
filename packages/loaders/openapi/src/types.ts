import { JSONSchemaLoaderOptions } from '@omnigraph/json-schema';

export interface OpenAPILoaderOptions extends Partial<JSONSchemaLoaderOptions> {
  // The URL or FileSystem path to the RAML API Document.
  oasFilePath: string;
}
