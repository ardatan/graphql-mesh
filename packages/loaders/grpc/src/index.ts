import type { Logger, YamlConfig } from "@graphql-mesh/types";
import { gRPCLoader } from "./gRPCLoader";

interface LoaderContext {
  cwd: string;
  logger: Logger;
}

export function loadGrpcSubgraph(name: string, options: YamlConfig.GrpcHandler) {
  return (ctx: LoaderContext) => ({
    name,
    schema$: new gRPCLoader(name, ctx.cwd, ctx.logger, options).buildSchema(),
  });
}
