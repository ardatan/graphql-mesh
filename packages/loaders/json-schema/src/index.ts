import { loadGraphQLSchemaFromJSONSchemas } from './loadGraphQLSchemaFromJSONSchemas.js';
export default loadGraphQLSchemaFromJSONSchemas;
export * from './loadGraphQLSchemaFromJSONSchemas.js';
export * from './getComposerFromJSONSchema.js';
export * from './getDereferencedJSONSchemaFromOperations.js';
export * from './getGraphQLSchemaFromDereferencedJSONSchema.js';
export * from './types.js';
export { processDirectives } from './directives.js';
export * from './bundle.js';
