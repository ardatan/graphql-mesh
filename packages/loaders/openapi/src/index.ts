export { loadGraphQLSchemaFromOpenAPI as default } from './loadGraphQLSchemaFromOpenAPI.js';
export * from './loadGraphQLSchemaFromOpenAPI.js';
export { getJSONSchemaOptionsFromOpenAPIOptions } from './getJSONSchemaOptionsFromOpenAPIOptions.js';
export { OpenAPILoaderOptions } from './types.js';
