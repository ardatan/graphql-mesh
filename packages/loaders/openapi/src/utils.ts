import { camelCase } from 'change-case';

export function getFieldNameFromPath(path: string, method: string, responseTypeSchemaRef: string) {
  // Replace identifiers with "by"
  path = path.split('{').join('by_').split('}').join('');

  const [actualPartsStr, allQueryPartsStr] = path.split('?');

  const actualParts = actualPartsStr.split('/').filter(Boolean);

  let fieldNameWithoutMethod = actualParts.join('_');

  // If path doesn't give any field name without identifiers, we can use the return type with HTTP Method name
  if ((!fieldNameWithoutMethod || fieldNameWithoutMethod.startsWith('by')) && responseTypeSchemaRef) {
    const refArr = responseTypeSchemaRef.split('/');
    // lowercase looks better in the schema
    const prefix = camelCase(refArr[refArr.length - 1]);
    if (fieldNameWithoutMethod) {
      fieldNameWithoutMethod = prefix + '_' + fieldNameWithoutMethod;
    } else {
      fieldNameWithoutMethod = prefix;
    }
  }

  if (allQueryPartsStr) {
    const queryParts = allQueryPartsStr.split('&');
    for (const queryPart of queryParts) {
      const [queryName] = queryPart.split('=');
      fieldNameWithoutMethod += '_' + 'by' + '_' + queryName;
    }
  }

  // get_ doesn't look good in field names
  if (method === 'get') {
    return fieldNameWithoutMethod;
  }
  return method + '_' + fieldNameWithoutMethod;
}
