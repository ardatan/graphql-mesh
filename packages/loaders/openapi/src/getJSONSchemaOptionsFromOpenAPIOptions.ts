import { getInterpolatedHeadersFactory, readFileOrUrl, sanitizeNameForGraphQL } from '@graphql-mesh/utils';
import { JSONSchemaObject } from 'json-machete';
import { OpenAPIV3, OpenAPIV2 } from 'openapi-types';
import { HTTPMethod, JSONSchemaHTTPJSONOperationConfig, JSONSchemaOperationConfig } from '@omnigraph/json-schema';
import { env } from 'process';
import { getFieldNameFromPath } from './utils';

interface GetJSONSchemaOptionsFromOpenAPIOptionsParams {
  oasFilePath: OpenAPIV3.Document | OpenAPIV2.Document | string;
  fallbackFormat?: 'json' | 'yaml' | 'js' | 'ts';
  cwd?: string;
  fetch?: WindowOrWorkerGlobalScope['fetch'];
  baseUrl?: string;
  schemaHeaders?: Record<string, string>;
  operationHeaders?: Record<string, string>;
  respectErrorResponses?: boolean;
}

export async function getJSONSchemaOptionsFromOpenAPIOptions({
  oasFilePath,
  fallbackFormat,
  cwd,
  fetch,
  baseUrl,
  schemaHeaders,
  operationHeaders,
  respectErrorResponses,
}: GetJSONSchemaOptionsFromOpenAPIOptionsParams) {
  const schemaHeadersFactory = getInterpolatedHeadersFactory(schemaHeaders);
  const oasOrSwagger: OpenAPIV3.Document | OpenAPIV2.Document =
    typeof oasFilePath === 'string'
      ? await readFileOrUrl(oasFilePath, {
          cwd,
          fallbackFormat,
          headers: schemaHeadersFactory({ env }),
          fetch,
        })
      : oasFilePath;
  const operations: JSONSchemaOperationConfig[] = [];

  if ('servers' in oasOrSwagger) {
    baseUrl = baseUrl || oasOrSwagger.servers[0].url;
  }

  for (const relativePath in oasOrSwagger.paths) {
    const pathObj = oasOrSwagger.paths[relativePath];
    for (const method in pathObj) {
      const methodObj = pathObj[method] as OpenAPIV2.OperationObject | OpenAPIV3.OperationObject;
      const operationConfig = {
        method: method.toUpperCase() as HTTPMethod,
        path: relativePath,
        type: method.toUpperCase() === 'GET' ? 'query' : 'mutation',
        field: methodObj.operationId,
        description: methodObj.description,
        schemaHeaders,
        operationHeaders,
      } as JSONSchemaHTTPJSONOperationConfig;
      operations.push(operationConfig);
      for (const paramObjIndex in methodObj.parameters) {
        const paramObj = methodObj.parameters[paramObjIndex] as OpenAPIV2.ParameterObject | OpenAPIV3.ParameterObject;
        switch (paramObj.in) {
          case 'query':
            if (method.toUpperCase() === 'GET') {
              const requestSchema = (operationConfig.requestSchema = operationConfig.requestSchema || {
                type: 'object',
                properties: {},
              }) as JSONSchemaObject;
              requestSchema.properties[paramObj.name] = paramObj.schema || paramObj;
              if (!requestSchema.properties[paramObj.name].title) {
                requestSchema.properties[paramObj.name].name = paramObj.name;
              }
              if (!requestSchema.properties[paramObj.name].description) {
                requestSchema.properties[paramObj.name].description = paramObj.description;
              }
            } else {
              if (!operationConfig.path.includes('?')) {
                operationConfig.path += '?';
              }
              operationConfig.path += `${paramObj.name}={args.${paramObj.name}}`;
              switch (paramObj.schema?.type || (paramObj as any).type) {
                case 'string':
                  operationConfig.argTypeMap = operationConfig.argTypeMap || {};
                  operationConfig.argTypeMap[paramObj.name] = 'String';
                  break;
                case 'integer':
                  operationConfig.argTypeMap = operationConfig.argTypeMap || {};
                  operationConfig.argTypeMap[paramObj.name] = 'Int';
                  break;
                case 'number':
                  operationConfig.argTypeMap = operationConfig.argTypeMap || {};
                  operationConfig.argTypeMap[paramObj.name] = 'Float';
                  break;
                case 'boolean':
                  operationConfig.argTypeMap = operationConfig.argTypeMap || {};
                  operationConfig.argTypeMap[paramObj.name] = 'Boolean';
                  break;
              }
            }
            break;
          case 'path':
            // If it is in the path, let JSON Schema handler put it
            operationConfig.path = operationConfig.path.replace(`{${paramObj.name}}`, `{args.${paramObj.name}}`);
            break;
          case 'header':
            operationConfig.headers = operationConfig.headers || {};
            operationConfig.headers[paramObj.name] = `{args.${paramObj.name}}`;
            break;
          case 'cookie': {
            operationConfig.headers = operationConfig.headers || {};
            operationConfig.headers.cookie = operationConfig.headers.cookie || '';
            const cookieParams = operationConfig.headers.cookie.split('; ');
            cookieParams.push(`${paramObj.name}={args.${paramObj.name}}`);
            operationConfig.headers.cookie = cookieParams.join('; ');
            break;
          }
          case 'body':
            operationConfig.requestSchema = `${oasFilePath}#/paths/${relativePath
              .split('/')
              .join('~1')}/${method}/parameters/${paramObjIndex}/schema`;
            break;
        }
      }

      if ('requestBody' in methodObj) {
        const requestBodyObj = methodObj.requestBody as OpenAPIV3.RequestBodyObject;
        const contentKey = Object.keys(requestBodyObj.content)[0];
        operationConfig.requestSchema = `${oasFilePath}#/paths/${relativePath
          .split('/')
          .join('~1')}/${method}/requestBody/content/${contentKey}/schema`;
      }

      const responseSchemaUnion = (operationConfig.responseSchema = {
        oneOf: [],
      }) as JSONSchemaObject;

      // Handling multiple response types
      for (const responseKey in methodObj.responses) {
        // Only take successful responses
        if (!respectErrorResponses && !(responseKey.startsWith('2') || responseKey === 'default')) {
          continue;
        }
        const responseObj = methodObj.responses[responseKey] as OpenAPIV3.ResponseObject | OpenAPIV2.ResponseObject;
        let schemaObj: JSONSchemaObject;

        if ('content' in responseObj) {
          const contentKey = Object.keys(responseObj.content)[0];
          schemaObj = responseObj.content[contentKey].schema as any;
          if (schemaObj) {
            responseSchemaUnion.oneOf.push({
              $ref: `${oasFilePath}#/paths/${relativePath
                .split('/')
                .join('~1')}/${method}/responses/${responseKey}/content/${contentKey}/schema`,
            });
          }
        } else if ('schema' in responseObj) {
          schemaObj = responseObj.schema as any;
          if (schemaObj) {
            responseSchemaUnion.oneOf.push({
              $ref: `${oasFilePath}#/paths/${relativePath
                .split('/')
                .join('~1')}/${method}/responses/${responseKey}/schema`,
            });
          }
        }

        if (!operationConfig.field) {
          // Operation ID might not be avaiable so let's generate field name from path and response type schema
          operationConfig.field = sanitizeNameForGraphQL(getFieldNameFromPath(relativePath, method, schemaObj?.$ref));
        }

        // Give a better name to the request input object
        if (typeof operationConfig.requestSchema === 'object' && !operationConfig.requestSchema.title) {
          operationConfig.requestSchema.title = operationConfig.field + '_request';
        }

        // If we don't need unsuccessful response types, just break the loop to have singular response type
        if (!respectErrorResponses) {
          break;
        }
      }

      // If response types are singular, no need to have unions
      if (responseSchemaUnion.oneOf.length <= 1) {
        operationConfig.responseSchema =
          (responseSchemaUnion.oneOf[0] as JSONSchemaObject)?.$ref || responseSchemaUnion.oneOf[0];
      }
    }
  }

  return {
    operations,
    baseUrl,
    cwd,
    fetch,
    schemaHeaders,
    operationHeaders,
  };
}
