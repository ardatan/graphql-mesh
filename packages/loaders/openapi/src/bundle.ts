import {
  getGraphQLSchemaFromBundle,
  createBundle as createJSONSchemaLoaderBundle,
  JSONSchemaLoaderBundle as OpenAPILoaderBundle,
} from '@omnigraph/json-schema';
import { getJSONSchemaOptionsFromOpenAPIOptions } from './getJSONSchemaOptionsFromOpenAPIOptions';
import { OpenAPILoaderOptions } from './types';

/**
 * Creates a bundle by downloading and resolving the internal references once
 * to load the schema locally later
 */
export async function createBundle(
  name: string,
  openApiLoaderOptions: OpenAPILoaderOptions
): Promise<OpenAPILoaderBundle> {
  const { operations, baseUrl, cwd, fetch, schemaHeaders, operationHeaders } =
    await getJSONSchemaOptionsFromOpenAPIOptions(openApiLoaderOptions);
  return createJSONSchemaLoaderBundle(name, {
    operations,
    baseUrl,
    cwd,
    fetch,
    schemaHeaders,
    operationHeaders,
  });
}

export { getGraphQLSchemaFromBundle, OpenAPILoaderBundle };
