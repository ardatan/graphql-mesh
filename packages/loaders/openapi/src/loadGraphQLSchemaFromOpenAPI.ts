import { loadGraphQLSchemaFromJSONSchemas } from '@omnigraph/json-schema';
import { OpenAPILoaderOptions } from '.';
import { getJSONSchemaOptionsFromOpenAPIOptions } from './getJSONSchemaOptionsFromOpenAPIOptions';

/**
 * Creates a local GraphQLSchema instance from a OpenAPI Document.
 * Everytime this function is called, the OpenAPI file and its dependencies will be resolved on runtime.
 * If you want to avoid this, use `createBundle` function to create a bundle once and save it to a storage
 * then load it with `loadGraphQLSchemaFromBundle`.
 */
export async function loadGraphQLSchemaFromOpenAPI(name: string, options: OpenAPILoaderOptions) {
  const extraJSONSchemaOptions = await getJSONSchemaOptionsFromOpenAPIOptions(options);
  return loadGraphQLSchemaFromJSONSchemas(name, {
    ...options,
    ...extraJSONSchemaOptions,
  });
}
