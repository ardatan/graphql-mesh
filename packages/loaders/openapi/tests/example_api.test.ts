import { execute, GraphQLObjectType, GraphQLSchema, parse } from 'graphql';
import { loadGraphQLSchemaFromOpenAPI } from '../src/loadGraphQLSchemaFromOpenAPI';

import { startServer, stopServer } from '../../../handlers/openapi/test/example_api_server';
import { join } from 'path';
import { fetch } from 'cross-undici-fetch';

let createdSchema: GraphQLSchema;
const PORT = 3002;
const baseUrl = `http://localhost:${PORT}/api`;

describe('example_api', () => {
  beforeAll(async () => {
    createdSchema = await loadGraphQLSchemaFromOpenAPI('test', {
      fetch,
      baseUrl,
      oasFilePath: join(__dirname, '../../../handlers/openapi/test/fixtures/example_oas.json'),
    });
    await startServer(PORT);
  });
  afterAll(() => stopServer());

  it('should get descriptions', async () => {
    // Get all the descriptions of the fields on the GraphQL object type car
    const carType = createdSchema.getType('car') as GraphQLObjectType;
    expect(carType).toBeDefined();
    const carFields = carType.getFields();
    expect(carFields).toBeDefined();
    expect(carFields.model.description).toBe('The model of the car.');
    expect(carFields.color.description).toBe('The color of the car.');
  });

  it('should get resource (incl. enum)', async () => {
    // Status is an enum
    const query = /* GraphQL */ `
      {
        getUserByUsername(username: "arlene") {
          name
          status
        }
      }
    `;

    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });

    expect(result).toEqual({
      data: { getUserByUsername: { name: 'Arlene L McMahon', status: 'staff' } },
    });
  });

  // OAS allows you to define response objects with HTTP code with the XX wildcard syntax
  it('should get resource with status code: 2XX', async () => {
    const query = /* GraphQL */ `
      {
        getPapers {
          name
          published
        }
      }
    `;

    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });
    expect(result).toEqual({
      data: {
        getPapers: [
          { name: 'Deliciousness of apples', published: true },
          { name: 'How much coffee is too much coffee?', published: false },
          {
            name: 'How many tennis balls can fit into the average building?',
            published: true,
          },
        ],
      },
    });
  });

  /**
   * Some operations do not have a response body.
   */
  it('should get resource with no response schema and status code: 204', async () => {
    const query = /* GraphQL */ `
      {
        getBonuses
      }
    `;

    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });
    expect(result).toEqual({
      data: {
        getBonuses: '',
      },
    });
  });

  // Link objects in the OAS allow OtG to create nested GraphQL objects that resolve on different API calls
  it('should get nested resource via link $response.body#/...', async () => {
    const query = /* GraphQL */ `
      {
        getUserByUsername(username: "arlene") {
          name
          employerCompany {
            legalForm
          }
        }
      }
    `;
    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });
    expect(result).toEqual({
      data: {
        getUserByUsername: {
          name: 'Arlene L McMahon',
          employerCompany: {
            legalForm: 'public',
          },
        },
      },
    });
  });

  it('should get nested resource via link $request.path#/... and $request.query#/', async () => {
    const query = /* GraphQL */ `
      {
        get_product_with_id(product_id: "123", input: { product_tag: "blah" }) {
          product_name
          reviews {
            text
          }
        }
      }
    `;
    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });
    expect(result).toEqual({
      data: {
        get_product_with_id: {
          product_name: 'Super Product',
          reviews: [{ text: 'Great product' }, { text: 'I love it' }],
        },
      },
    });
  });

  // Both an operationId and an operationRef can be used to create a link object
  it('should get nested resource via link operationRef', async () => {
    const query = /* GraphQL */ `
      {
        get_product_with_id(product_id: "123", input: { product_tag: "blah" }) {
          product_name
          reviewsWithOperationRef {
            text
          }
        }
      }
    `;
    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });
    expect(result).toEqual({
      data: {
        get_product_with_id: {
          product_name: 'Super Product',
          reviewsWithOperationRef: [{ text: 'Great product' }, { text: 'I love it' }],
        },
      },
    });
  });

  it('should get nested lists of resources', async () => {
    const query = /* GraphQL */ `
      {
        getUserByUsername(username: "arlene") {
          name
          friends {
            name
            friends {
              name
              friends {
                name
              }
            }
          }
        }
      }
    `;
    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });
    expect(result).toEqual({
      data: {
        getUserByUsername: {
          name: 'Arlene L McMahon',
          friends: [
            {
              name: 'William B Ropp',
              friends: [
                {
                  name: 'William B Ropp',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
                {
                  name: 'John C Barnes',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
                {
                  name: 'Heather J Tate',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
              ],
            },
            {
              name: 'John C Barnes',
              friends: [
                {
                  name: 'William B Ropp',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
                {
                  name: 'John C Barnes',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
                {
                  name: 'Heather J Tate',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
              ],
            },
            {
              name: 'Heather J Tate',
              friends: [
                {
                  name: 'William B Ropp',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
                {
                  name: 'John C Barnes',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
                {
                  name: 'Heather J Tate',
                  friends: [
                    {
                      name: 'William B Ropp',
                    },
                    {
                      name: 'John C Barnes',
                    },
                    {
                      name: 'Heather J Tate',
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    });
  });

  it('should get nested lists of resources without specifying a path param for the parent resource', async () => {
    const query = /* GraphQL */ `
      {
        getUsers(input: { limit: 1 }) {
          name
          friends {
            name
            friends {
              name
              friends {
                name
              }
            }
          }
        }
      }
    `;
    const result = await execute({
      schema: createdSchema,
      document: parse(query),
    });

    expect(result).toEqual({
      data: {
        getUsers: [
          {
            name: 'Arlene L McMahon',
            friends: [
              {
                name: 'William B Ropp',
                friends: [
                  {
                    name: 'William B Ropp',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                  {
                    name: 'John C Barnes',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                  {
                    name: 'Heather J Tate',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                ],
              },
              {
                name: 'John C Barnes',
                friends: [
                  {
                    name: 'William B Ropp',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                  {
                    name: 'John C Barnes',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                  {
                    name: 'Heather J Tate',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                ],
              },
              {
                name: 'Heather J Tate',
                friends: [
                  {
                    name: 'William B Ropp',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                  {
                    name: 'John C Barnes',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                  {
                    name: 'Heather J Tate',
                    friends: [
                      {
                        name: 'William B Ropp',
                      },
                      {
                        name: 'John C Barnes',
                      },
                      {
                        name: 'Heather J Tate',
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    });
  });
});
