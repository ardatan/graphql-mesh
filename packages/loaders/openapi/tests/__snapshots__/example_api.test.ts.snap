// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example_api should generate the schema correctly 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

directive @oneOf on INPUT_OBJECT | FIELD_DEFINITION

type Query {
  \\"Return a list of users.\\"
  getUsers(input: getUsers_request_Input!): [user]
  \\"Returns a user from the system.\\"
  getUserByUsername(username: String!): user
  \\"Returns a car to test nesting of sub operations\\"
  getUserCar(username: String!): car
  \\"Returns a user's friends to test pagination\\"
  getUserFriends(username: String!): [user]
  \\"Returns all company assets\\"
  getAllAssets(companyId: String!): [query_getAllAssets_items]
  \\"Returns information about all employee cars\\"
  getAllCars: [car]
  \\"Returns the profile of a company with the given ID.\\"
  getCompanyById(id: String!): company
  \\"Used to test cookies.\\"
  getCookie(cookie_type: String!, cookie_size: String!): String
  \\"Used to find the nearest coffee machine based on the user's coordinates. Used to test the content field in parameter objects.\\"
  getNearestCoffeeMachine(input: getNearestCoffeeMachine_request_Input): coordinates
  \\"Used to test link parameters with variables\\"
  getCopier(input: getCopier_request_Input): getCopier_200_response
  \\"Used to test generation of object types with matching schema.\\"
  getNumberOfCleanDesks: String
  \\"Used to test generation of object types with matching schema.\\"
  getNumberOfDirtyDesks: String
  \\"Used to test empty responses\\"
  getBonuses: Void
  \\"Return an office.\\"
  getOffice(accept: String, id: Int!): office
  \\"Return a list of papers. Endpoint to test 2XX status code.\\"
  getPapers: [paper]
  \\"Another endpoint to test unsanitized parameters and data.\\"
  get_product_with_id(input: get_product_with_id_request_Input!, product_id: String!): product_with_id
  \\"Obtain reviews for a product.\\"
  getProductReviews(input: getProductReviews_request_Input!, id: String!): [query_getProductReviews_items]
  \\"An endpoint to test authentication.\\"
  get_patent_with_id(patent_id: String!): patent_with_id
  \\"Another endpoint to test authentication.\\"
  get_project_with_id(project_id: Int!): project_with_id
  \\"Used to test link parameters with variables\\"
  getScanner(input: getCopier_request_Input): getCopier_200_response
  \\"Endpoint to test sending of headers.\\"
  getSnack(snack_type: String!, snack_size: String!): String
  \\"Endpoint to test sending of options.\\"
  get_Status(input: get_Status_request_Input!): String
  \\"Used to test OAuth token being present in header.\\"
  getSecure: String
  \\"Returns the (contents of a) trashcan from a specific office\\"
  getAllTrashCans: [trashcan]
  \\"Returns the (contents of a) trashcan from a specific owner\\"
  getOfficeTrashCan(username: String!): trashcan
  \\"It could be anything!\\"
  mystery: mystery_200_response
}

\\"A user represents a natural person\\"
type user {
  \\"The identifier of a user\\"
  username: String
  \\"The legal name of a user\\"
  name: String!
  address: address!
  address2: address
  \\"The identifier of the company a user works for\\"
  employerId: String!
  \\"The hobbies of this user\\"
  hobbies: [String]!
  status: query_getUsers_items_status
  nomenclature: query_getUsers_items_nomenclature
  \\"Allows to fetch the user's employer company.\\"
  employerCompany: company
  \\"Allows to fetch the user's friends.\\"
  friends: [user]
}

\\"The legal address of a user\\"
type address {
  \\"The street a user lives in\\"
  street: String
  \\"The city a user lives in\\"
  city: String
}

enum query_getUsers_items_status {
  staff
  contractor
  alumni
}

type query_getUsers_items_nomenclature {
  species: String
  suborder: String
  family: String!
  familyCircular: familyObject
  genus: String
}

type familyObject {
  family: String!
  familyCircular: familyObject
}

\\"A company is the employer of many users\\"
type company {
  \\"The identifier of a company\\"
  id: String
  \\"The legal name of a company\\"
  name: String
  \\"The legal form of a company\\"
  legalForm: String
  \\"The identifier of the user who is CEO of the company\\"
  ceoUsername: String
  \\"The offices of a company\\"
  offices: [address]
  \\"Allows to fetch the company's CEO user.\\"
  ceoUser: user
}

input getUsers_request_Input {
  \\"Limit of the number of users to return.\\"
  limit: Int!
}

\\"A car\\"
type car {
  \\"The model of the car.\\"
  model: String
  \\"The color of the car.\\"
  color: String
  features: query_getUserCar_features
  tags: Tags
  kind: query_getUserCar_kind
  rating: query_getUserCar_rating
}

type query_getUserCar_features {
  additionalProperties: JSON
}

\\"The \`JSON\` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).\\"
scalar JSON @specifiedBy(url: \\"http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf\\")

\\"Arbitrary (string) tags describing an entity.\\"
type Tags {
  additionalProperties: JSON
}

enum query_getUserCar_kind {
  SEDAN
  SUV
  MINIVAN
  LIMOSINE
  RACE_CAR
}

\\"The rating of the car.\\"
enum query_getUserCar_rating {
  _100
  _101
  _200
}

union query_getAllAssets_items = car | user | trashcan

type trashcan {
  brand: query_getAllAssets_items_oneOf_2_brand
  contents: [query_getAllAssets_items_oneOf_2_brand]
}

scalar query_getAllAssets_items_oneOf_2_brand

type coordinates {
  lat: Float!
  long: Float!
}

input getNearestCoffeeMachine_request_Input {
  lat: Float
  long: Float
}

type getCopier_200_response {
  body: String
  \\"Used to test link parameters with variables\\"
  basicLink: getCopier_200_response
  \\"Used to test link parameters with variables\\"
  constantLink: getCopier_200_response
  \\"Used to test link parameters with variables\\"
  variableLink: getCopier_200_response
  \\"Used to test link parameters with variables\\"
  everythingLink: getCopier_200_response
  \\"Used to test link parameters with variables\\"
  everythingLink2: getCopier_200_response
}

input getCopier_request_Input {
  query: String
}

\\"Represents empty values\\"
scalar Void

\\"An office\\"
type office {
  \\"The owner ID of the office\\"
  employeeId: String
  \\"The room number of the office\\"
  room_number: Int
  \\"The identifier of the company a user works for\\"
  employerId: String
  \\"Allows to fetch the user who invented a product.\\"
  owner: user
  \\"Allows to fetch the user's employer company.\\"
  company: company
}

\\"A research paper\\"
type paper {
  \\"The name of the paper\\"
  name: String
  \\"Whether the paper is published\\"
  published: Boolean
}

\\"A product\\"
type product_with_id {
  \\"The name of the product\\"
  product_name: String
  \\"The id of the product\\"
  product_id: String
  \\"A tag associated with the product\\"
  product_tag: String
  \\"Obtain reviews for a product.\\"
  reviews: [query_getProductReviews_items]
  \\"Obtain reviews for a product.\\"
  reviewsWithOperationRef: [query_getProductReviews_items]
}

type query_getProductReviews_items {
  text: String
  timestamp: BigInt
}

\\"The \`BigInt\` scalar type represents non-fractional signed whole numeric values.\\"
scalar BigInt

input get_product_with_id_request_Input {
  product_tag: String!
}

input getProductReviews_request_Input {
  product_tag: String!
}

\\"A patent\\"
type patent_with_id {
  \\"The id of the patent\\"
  patent_id: String!
  \\"The id of the inventor\\"
  inventor_id: String
  \\"Allows to fetch the user who invented a product.\\"
  inventor: user
}

\\"A project\\"
type project_with_id {
  \\"The id of the project\\"
  project_id: Int
  \\"The id of the lead\\"
  lead_id: String
  \\"Whether the project is active\\"
  active: Boolean
  \\"The patent associated with a project if applicable\\"
  patentId: String
  \\"Allows to fetch the user leading a project.\\"
  projectLead: user
  \\"Allows to fetch a patent.\\"
  patent: patent_with_id
  \\"Allows to fetch the user leading a project.\\"
  ProjectLead: user
}

input get_Status_request_Input {
  globalquery: String!
  limit: Int!
}

union mystery_200_response = query_mystery_oneOf_0 | query_mystery_oneOf_1

type query_mystery_oneOf_0 {
  common_attribute: String
}

type query_mystery_oneOf_1 {
  different_attribute: String
}

type Mutation {
  \\"Create a new user in the system.\\"
  postUser(input: user_Input!): user
  \\"Create a new paper in the system. Endpoint to test non-application/json request and response bodies.\\"
  postPaper(input: String): String
  \\"Endpoint to test unsanitized parameters and data.\\"
  post_product_with_id(input: product_with_id_Input): product_with_id
  \\"An endpoint to test authenticated POST requests.\\"
  post_project_with_id(input: project_with_id_Input): project_with_id
  \\"Used to test link parameters with variables\\"
  postScanner(input: String, path: String!, query: String): getCopier_200_response
  \\"Endpoint to test placeholder objects to wrap response objects.\\"
  post_status(input: paths_status_post_requestBody_content_application_json_schema_Input, globalquery: String!): String
  \\"Add new contents to the trashcan of a specific owner\\"
  postOfficeTrashCan(input: paths_trashcans__LEFT_CURLY_BRACE_username_RIGHT_CURLY_BRACE__post_requestBody_content_application_json_schema_Input, username: String!): trashcan
}

\\"A user represents a natural person\\"
input user_Input {
  \\"The identifier of a user\\"
  username: String
  \\"The legal name of a user\\"
  name: String!
  address: address_Input!
  address2: address_Input
  \\"The identifier of the company a user works for\\"
  employerId: String!
  \\"The hobbies of this user\\"
  hobbies: [String]!
  status: query_getUsers_items_status
  nomenclature: query_getUsers_items_nomenclature_Input
}

\\"The legal address of a user\\"
input address_Input {
  \\"The street a user lives in\\"
  street: String
  \\"The city a user lives in\\"
  city: String
}

input query_getUsers_items_nomenclature_Input {
  species: String
  suborder: String
  family: String!
  familyCircular: familyObject_Input
  genus: String
}

input familyObject_Input {
  family: String!
  familyCircular: familyObject_Input
}

\\"A product\\"
input product_with_id_Input {
  \\"The name of the product\\"
  product_name: String
  \\"The id of the product\\"
  product_id: String
  \\"A tag associated with the product\\"
  product_tag: String
}

\\"A project\\"
input project_with_id_Input {
  \\"The id of the project\\"
  project_id: Int
  \\"The id of the lead\\"
  lead_id: String
  \\"Whether the project is active\\"
  active: Boolean
  \\"The patent associated with a project if applicable\\"
  patentId: String
}

input paths_status_post_requestBody_content_application_json_schema_Input {
  hello: String
}

scalar paths_trashcans__LEFT_CURLY_BRACE_username_RIGHT_CURLY_BRACE__post_requestBody_content_application_json_schema_Input"
`;
